semgrep-core/src/fixing/Autofix_metavar_replacement.ml

```diff
@@ -0,0 +1,125 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open AST_generic
+module MV = Metavariable
+
+(******************************************************************************)
+(* Module responsible for traversing a fix pattern AST and replacing
+ * metavariables within it with the AST nodes from the target file to which the
+ * metavariables are bound. *)
+(******************************************************************************)
+
+let make_metavar_tbl bindings =
+  let tbl = Hashtbl.create (List.length bindings) in
+  List.iter (fun (mvar, mvalue) -> Hashtbl.replace tbl mvar mvalue) bindings;
+  tbl
+
+let replace metavar_tbl pattern_ast =
+  (* Use a mapper to traverse the AST. For each metavar, look up what it is
+   * bound to. If the kind of node matches, replace the metavar in the pattern
+   * with the AST node to which it is bound. Note that we can't handle this by
+   * simply visiting the identifier nodes, because metavars are often bound to
```

Comment:
I thought I understood this comment and then I didn't. The way I initially read the comment was "because metavars can exist in constructs other than Id", which I then realized (from looking at `has_remaining_metavars`) wasn't true. Now I think it's because by the time you get to `Id` you've lost context that you need earlier. This would probably be obvious to me if I thought about it, but thinking is hard

Reply:
Basically, you can't replace an `ident` type with an `argument` type. If you want to replace arguments, you have to start with arguments. I think an example might help drive the point home, so I'll add that to this comment.